python/test/unit/fem/test_custom_assembler.py

```diff
@@ -269,6 +269,97 @@ def assemble_matrix_ctypes(A, mesh, dofmap, set_vals, mode):
         set_vals(A, 3, rows.ctypes, 3, cols.ctypes, A_local.ctypes, mode)
 
 
+@numba.njit
+def assemble_exterior_facets_cffi(A, kernel, mesh, coeffs, consts, perm, dofmap,
+                                  num_dofs_per_element, gdim, facet_info, set_vals, mode):
+    """Assemble exterior facet integral into a PETSc matrix A"""
+
+    # Mesh data
+    pos, x_dofmap, x = mesh
+    geometry = np.zeros((pos[1] - pos[0], gdim))
+
+    facet_index = np.zeros(1, dtype=np.int32)
+    facet_perm = np.zeros(1, dtype=np.uint8)
+
+    A_local = np.zeros((num_dofs_per_element, num_dofs_per_element),
+                       dtype=PETSc.ScalarType)
+    # Permutation info
+    cell_perms, facet_perms = perm
+    for i in range(facet_info.shape[0]):
+        cell_index, local_facet = facet_info[i]
+        cell = pos[cell_index]
+        facet_index[0] = local_facet
+        num_vertices = pos[cell_index + 1] - pos[cell_index]
+        # FIXME: This assumes a particular geometry dof layout
+        c = x_dofmap[cell:cell + num_vertices]
+        for j in range(num_vertices):
+            for k in range(gdim):
+                geometry[j, k] = x[c[j], k]
+        A_local.fill(0.0)
+        facet_perm[0] = facet_perms[local_facet, cell_index]
+        kernel(ffi.from_buffer(A_local),
+               ffi.from_buffer(coeffs[cell_index, :]),
+               ffi.from_buffer(consts),
+               ffi.from_buffer(geometry),
+               ffi.from_buffer(facet_index),
+               ffi.from_buffer(facet_perm),
+               cell_perms[cell_index])
+
+        local_pos = dofmap[num_dofs_per_element * cell_index:
+                           num_dofs_per_element * cell_index
+                           + num_dofs_per_element]
+
+        ierr_loc = set_vals(A, num_dofs_per_element, ffi.from_buffer(local_pos),
+                            num_dofs_per_element, ffi.from_buffer(local_pos),
+                            ffi.from_buffer(A_local), mode)
+        if ierr_loc != 0:
+            raise ValueError("Assembly failed")
+    sink(A_local, local_pos)
+
+
+def pack_facet_info(mesh, integrals, i):
```

Comment:
Seem more appropriate as user tests - not really testing core DOLFIN functionality.